### PR TITLE
UCT/IB/SHM: portion of fixes for PGI compiler

### PR DESCRIPTION
**Pull request by hoopoepg**
- added atomic_sub call
- fixed void* arithmetics
- minor datatype conversions
- suppressed compiler error on dynamic size arrays

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/12361/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
<details><summary>Mellanox CI: <b>PASSED</b#  > on 29 workers (click for details)</summary>

**Note:** the logs will be deleted after 24-Sep-2019

Agent/Stage | Status
-- | --
[_main](https://gist.githubusercontent.com/mellanox-github/8b6f415ad7e49eacc719726aba50a2ad/raw/5cc1861751015afe146295d48994d098a312071b/_main) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/8b6f415ad7e49eacc719726aba50a2ad/raw/1b529dfdb37fd24a793ce690625d77d6fe74e809/hpc-arm-cavium-jenkins_W0) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/8b6f415ad7e49eacc719726aba50a2ad/raw/426c3a95325ca70324c8fe1f52b0595b2e3fe200/hpc-arm-cavium-jenkins_W1) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/8b6f415ad7e49eacc719726aba50a2ad/raw/4ae6d65b5978127b501fa93f83627df597bdd58f/hpc-arm-cavium-jenkins_W2) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/8b6f415ad7e49eacc719726aba50a2ad/raw/1576ee152184cb14e5dd1f778cbecf894dc34fa4/hpc-arm-cavium-jenkins_W3) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/8b6f415ad7e49eacc719726aba50a2ad/raw/d56e370a1e94adb855569d8d1afcae66e94b0e02/hpc-arm-hwi-jenkins_W0) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/8b6f415ad7e49eacc719726aba50a2ad/raw/10b3a13b3d89a2627832c2ccf56a92e6757bb589/hpc-arm-hwi-jenkins_W1) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/8b6f415ad7e49eacc719726aba50a2ad/raw/7868e9c11770b00edd251fb5faabb0f87c512dc3/hpc-arm-hwi-jenkins_W2) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/8b6f415ad7e49eacc719726aba50a2ad/raw/423eb714ecd276cfd898e4cd1ba5849937700857/hpc-arm-hwi-jenkins_W3) | :heavy_check_mark: SUCCESS
[hpc-test-althca_W0](https://gist.githubusercontent.com/mellanox-github/8b6f415ad7e49eacc719726aba50a2ad/raw/7a9914700665ae66ef239cd0a4080d35c53013c8/hpc-test-althca_W0) | :heavy_check_mark: SUCCESS
[hpc-test-althca_W1](https://gist.githubusercontent.com/mellanox-github/8b6f415ad7e49eacc719726aba50a2ad/raw/4767eb5c43754cefcd821e690427fcb71d162c7a/hpc-test-althca_W1) | :heavy_check_mark: SUCCESS
[hpc-test-althca_W2](https://gist.githubusercontent.com/mellanox-github/8b6f415ad7e49eacc719726aba50a2ad/raw/36aabd6a39d162122b2850cdcc79025008b7e237/hpc-test-althca_W2) | :heavy_check_mark: SUCCESS
[hpc-test-althca_W3](https://gist.githubusercontent.com/mellanox-github/8b6f415ad7e49eacc719726aba50a2ad/raw/5cf111e05b815f778460a1c2eb5b0bcfc01516db/hpc-test-althca_W3) | :heavy_check_mark: SUCCESS
[hpc-test-node-gpu_W0](https://gist.githubusercontent.com/mellanox-github/8b6f415ad7e49eacc719726aba50a2ad/raw/462745a6ecc996619b828cc54bb618dec4d60571/hpc-test-node-gpu_W0) | :heavy_check_mark: SUCCESS
[hpc-test-node-gpu_W1](https://gist.githubusercontent.com/mellanox-github/8b6f415ad7e49eacc719726aba50a2ad/raw/802d7935d9837560f7d79631115d9d52eee48a93/hpc-test-node-gpu_W1) | :heavy_check_mark: SUCCESS
[hpc-test-node-gpu_W2](https://gist.githubusercontent.com/mellanox-github/8b6f415ad7e49eacc719726aba50a2ad/raw/41f2cee2f1f48a462e80e9499098f05ec44d4f74/hpc-test-node-gpu_W2) | :heavy_check_mark: SUCCESS
[hpc-test-node-gpu_W3](https://gist.githubusercontent.com/mellanox-github/8b6f415ad7e49eacc719726aba50a2ad/raw/dd5bb8957b525cd418f5662a56949c4ada3e5b41/hpc-test-node-gpu_W3) | :heavy_check_mark: SUCCESS
[hpc-test-node-legacy_W0](https://gist.githubusercontent.com/mellanox-github/8b6f415ad7e49eacc719726aba50a2ad/raw/9888f045f1e06c38c33e9688f83d15d620df636a/hpc-test-node-legacy_W0) | :heavy_check_mark: SUCCESS
[hpc-test-node-legacy_W1](https://gist.githubusercontent.com/mellanox-github/8b6f415ad7e49eacc719726aba50a2ad/raw/a6326983f2e91ae6321cd223a8f78643fba20d2c/hpc-test-node-legacy_W1) | :heavy_check_mark: SUCCESS
[hpc-test-node-legacy_W2](https://gist.githubusercontent.com/mellanox-github/8b6f415ad7e49eacc719726aba50a2ad/raw/5758862e771da2ea74d61eeaf833e8715d695e2a/hpc-test-node-legacy_W2) | :heavy_check_mark: SUCCESS
[hpc-test-node-legacy_W3](https://gist.githubusercontent.com/mellanox-github/8b6f415ad7e49eacc719726aba50a2ad/raw/3f919726a6809199541a6f2bfb1efa2067f95199/hpc-test-node-legacy_W3) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W0](https://gist.githubusercontent.com/mellanox-github/8b6f415ad7e49eacc719726aba50a2ad/raw/eb2408154ea15fd9a91a1f0c8a854318bc0044da/hpc-test-node-new_W0) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W1](https://gist.githubusercontent.com/mellanox-github/8b6f415ad7e49eacc719726aba50a2ad/raw/855d77e5fb5f6f746bf48e005c1a3ad8e8525653/hpc-test-node-new_W1) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W2](https://gist.githubusercontent.com/mellanox-github/8b6f415ad7e49eacc719726aba50a2ad/raw/fdace5829632fc9a73a955b079a0c770b3dcb680/hpc-test-node-new_W2) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W3](https://gist.githubusercontent.com/mellanox-github/8b6f415ad7e49eacc719726aba50a2ad/raw/541a92c3e5e2c2998af2a169e73e540a6692c684/hpc-test-node-new_W3) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/8b6f415ad7e49eacc719726aba50a2ad/raw/28e45a094a3a607b1a5f84b35733db728635b09b/r-vmb-ppc-jenkins_W0) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/8b6f415ad7e49eacc719726aba50a2ad/raw/b59fd31fbd2301bbfce121cc55e5a2fcd9e5eb67/r-vmb-ppc-jenkins_W1) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/8b6f415ad7e49eacc719726aba50a2ad/raw/b2c3889aac1a62d942ed9d960a492e574e5f4fa4/r-vmb-ppc-jenkins_W2) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/8b6f415ad7e49eacc719726aba50a2ad/raw/3b59c06dc1cf4cf1d9cecb806f4ac702be4258c5/r-vmb-ppc-jenkins_W3) | :heavy_check_mark: SUCCESS

**Comment by hoopoepg** [comment]
bot:pipe:retest

File: src/uct/base/uct_md.h [hunk 1]
```diff
@@ -32,34 +32,59 @@ struct uct_md_config {
 };
 
 
-/**
- * Memory domain operations
- */
-struct uct_md_ops {
-    void         (*close)(uct_md_h md);
+    typedef void (*uct_md_close_func_t)(uct_md_h md);
```
Comment (yosefe) [inline src/uct/base/uct_md.h]: why do we have indentation here?

File: src/uct/base/uct_md.h [hunk 2]
```diff
@@ -32,34 +32,59 @@ struct uct_md_config {
...
+                                                    size_t *length_p,
+                                                    void **address_p,
+                                                    unsigned flags,
+                                                    const char *alloc_name,
+                                                    uct_mem_h *memh_p);
 
-    ucs_status_t (*mem_alloc)(uct_md_h md, size_t *length_p, void **address_p,
-                              unsigned flags, const char *alloc_name,
-                              uct_mem_h *memh_p);
+    typedef ucs_status_t (*uct_md_mem_free_func_t)(uct_md_h md, uct_mem_h memh);
```
Comment (yosefe) [inline src/uct/base/uct_md.h]: space line after

File: src/uct/base/uct_md.h (hunk 1 above)
Comment (hoopoepg) [reply src/uct/base/uct_md.h]: wanted to put all typedefs into separate structure to eliminate prefix in datatypes... doesn't work :)

indentation is fixed

File: src/uct/base/uct_md.h (hunk 2 above)
Comment (hoopoepg) [reply src/uct/base/uct_md.h]: yep, missed

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/12399/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
<details><summary>Mellanox CI: <b>PASSED</b#  > on 29 workers (click for details)</summary>

**Note:** the logs will be deleted after 30-Sep-2019

Agent/Stage | Status
-- | --
[_main](https://gist.githubusercontent.com/mellanox-github/4229c2ac9ac879949137b203f7330345/raw/9c3ca053535d0e4a167f1bde0c5abedabed31fc5/_main) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/4229c2ac9ac879949137b203f7330345/raw/8e1bd24acd1bddd679e43e9d6164eacc4fe54436/hpc-arm-cavium-jenkins_W0) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/4229c2ac9ac879949137b203f7330345/raw/2df92245f14518526fb0fda9d7fc7f5bbe9f708d/hpc-arm-cavium-jenkins_W1) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/4229c2ac9ac879949137b203f7330345/raw/28882c72ff0c359132bb35eeff5492ed28c8eb9b/hpc-arm-cavium-jenkins_W2) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/4229c2ac9ac879949137b203f7330345/raw/074584f884b601a2dc9fc9a1f7861c84502ffcba/hpc-arm-cavium-jenkins_W3) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/4229c2ac9ac879949137b203f7330345/raw/9ad19bd96cf58541d1b6e27642c94606eac10339/hpc-arm-hwi-jenkins_W0) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/4229c2ac9ac879949137b203f7330345/raw/8c6ac553163844990fca88dd22068e839042f949/hpc-arm-hwi-jenkins_W1) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/4229c2ac9ac879949137b203f7330345/raw/eef3116531c8a96e2901fcef772649aa1423583f/hpc-arm-hwi-jenkins_W2) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/4229c2ac9ac879949137b203f7330345/raw/84c2c703222f4dd7bfa0c37a2e8c651661e5d6e2/hpc-arm-hwi-jenkins_W3) | :heavy_check_mark: SUCCESS
[hpc-test-althca_W0](https://gist.githubusercontent.com/mellanox-github/4229c2ac9ac879949137b203f7330345/raw/247bd508f48b8ad56f9d9be30d6b822dc0590993/hpc-test-althca_W0) | :heavy_check_mark: SUCCESS
[hpc-test-althca_W1](https://gist.githubusercontent.com/mellanox-github/4229c2ac9ac879949137b203f7330345/raw/d33041340d83b1864eea1f794bc6dfadb3c386f5/hpc-test-althca_W1) | :heavy_check_mark: SUCCESS
[hpc-test-althca_W2](https://gist.githubusercontent.com/mellanox-github/4229c2ac9ac879949137b203f7330345/raw/8e1d24125c298ba3fc7c1e2c46df039520f2cc00/hpc-test-althca_W2) | :heavy_check_mark: SUCCESS
[hpc-test-althca_W3](https://gist.githubusercontent.com/mellanox-github/4229c2ac9ac879949137b203f7330345/raw/85cb2e330cc5764fc24bb53b02ed50655337db2d/hpc-test-althca_W3) | :heavy_check_mark: SUCCESS
[hpc-test-node-gpu_W0](https://gist.githubusercontent.com/mellanox-github/4229c2ac9ac879949137b203f7330345/raw/652603cd7c0d24e6b75db3036664ec0ab33254e5/hpc-test-node-gpu_W0) | :heavy_check_mark: SUCCESS
[hpc-test-node-gpu_W1](https://gist.githubusercontent.com/mellanox-github/4229c2ac9ac879949137b203f7330345/raw/035a9a07320ca256cdb76650df75cb6d0a1c1362/hpc-test-node-gpu_W1) | :heavy_check_mark: SUCCESS
[hpc-test-node-gpu_W2](https://gist.githubusercontent.com/mellanox-github/4229c2ac9ac879949137b203f7330345/raw/5ff48f22acfbf7e56f3b75d87a3efcb1df7d3872/hpc-test-node-gpu_W2) | :heavy_check_mark: SUCCESS
[hpc-test-node-gpu_W3](https://gist.githubusercontent.com/mellanox-github/4229c2ac9ac879949137b203f7330345/raw/dba4c5fb532864c2561aa929251e8fd051757f79/hpc-test-node-gpu_W3) | :heavy_check_mark: SUCCESS
[hpc-test-node-legacy_W0](https://gist.githubusercontent.com/mellanox-github/4229c2ac9ac879949137b203f7330345/raw/0b398f8bbb4701e2c698cfb6bc9e61732e32faa5/hpc-test-node-legacy_W0) | :heavy_check_mark: SUCCESS
[hpc-test-node-legacy_W1](https://gist.githubusercontent.com/mellanox-github/4229c2ac9ac879949137b203f7330345/raw/c294f8ff0087736487e6a00d8c57b94d3815b8d4/hpc-test-node-legacy_W1) | :heavy_check_mark: SUCCESS
[hpc-test-node-legacy_W2](https://gist.githubusercontent.com/mellanox-github/4229c2ac9ac879949137b203f7330345/raw/847924020abc5969a6538bcf68c81d65c862dced/hpc-test-node-legacy_W2) | :heavy_check_mark: SUCCESS
[hpc-test-node-legacy_W3](https://gist.githubusercontent.com/mellanox-github/4229c2ac9ac879949137b203f7330345/raw/ad037a87e8c87c831d97666af57bd57360258d6d/hpc-test-node-legacy_W3) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W0](https://gist.githubusercontent.com/mellanox-github/4229c2ac9ac879949137b203f7330345/raw/1f22eb88fd91e4a2569cd62c6060c27fe04c8443/hpc-test-node-new_W0) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W1](https://gist.githubusercontent.com/mellanox-github/4229c2ac9ac879949137b203f7330345/raw/36ffd0953fee3bb1ff90d4f09fdd97b296343d01/hpc-test-node-new_W1) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W2](https://gist.githubusercontent.com/mellanox-github/4229c2ac9ac879949137b203f7330345/raw/069d6be005d94db46afa2fec911f42f69cf08d96/hpc-test-node-new_W2) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W3](https://gist.githubusercontent.com/mellanox-github/4229c2ac9ac879949137b203f7330345/raw/dd55bb082177d7aaa29d7fdb81f4a0a2ec973e93/hpc-test-node-new_W3) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/4229c2ac9ac879949137b203f7330345/raw/e2640d489a919fa71166da23597b4e0528ed498e/r-vmb-ppc-jenkins_W0) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/4229c2ac9ac879949137b203f7330345/raw/a169b9d4754366ac97cedabcc7e82c047a5f1e55/r-vmb-ppc-jenkins_W1) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/4229c2ac9ac879949137b203f7330345/raw/301f782ee0116b4094f88c8b36328f12a7182ccd/r-vmb-ppc-jenkins_W2) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/4229c2ac9ac879949137b203f7330345/raw/a1aa5ae47be6cad63868e009844c3e651783aa21/r-vmb-ppc-jenkins_W3) | :heavy_check_mark: SUCCESS

**Pull request merged**